02-peer-protocol.md

```diff
@@ -49,11 +68,464 @@ transaction is confirmed are also not persistent - until you know the script
+contributions.
+
+1. type: 70 (`tx_complete`)
+2. data:
+    * [`channel_id`:`channel_id`]
+
+#### Requirements
+
+The nodes:
+  - MUST send this message in succession to conclude this protocol
+
+The receiving node:
+  - MUST use the negotiated inputs and outputs to construct a transaction
+  - MUST fail the negotiation if:
+    - the peer's total input satoshis is less than their outputs
+    - the peer's paid feerate does not meet or exceed the agreed `feerate`,
+      (based on the `minimum fee`).
+    - if is the *non-initiator*:
+      - the *initiator*'s fees do not cover the `common` fields
+    - there are more than 252 inputs
+    - there are more than 252 outputs
+    - the estimated weight of the tx is greater than 400,000 (`MAX_STANDARD_TX_WEIGHT`)
+
+#### Rationale
+
+To signal the conclusion of exchange of transaction inputs and outputs.
+
+Upon successful exchange of `tx_complete` messages, both nodes
+should construct the transaction and proceed to the next portion of the
+protocol. For channel establishment v2, exchanging commitment transactions.
+
+For the `minimum fee` calculation see [BOLT #3](03-transactions.md#calculating-fees-for-collaborative-transaction-construction).
+
+The maximum inputs and outputs are capped at 252. This effectively fixes
+the byte size of the input and output counts on the transaction to one (1).
+
+### The `tx_signatures` Message
+
+1. type: 71 (`tx_signatures`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`sha256`:`txid`]
+    * [`u16`:`num_witnesses`]
+    * [`num_witnesses*witness_stack`:`witness_stack`]
+
+1. subtype: `witness_stack`
+2. data:
+    * [`u16`:`num_input_witness`]
+    * [`num_input_witness*witness_element`:`witness_element`]
+
+1. subtype: `witness_element`
+2. data:
+    * [`u16`:`len`]
+    * [`len*byte`:`witness`]
+
+#### Requirements
+
+The sending node:
+  - if it has the lowest total satoshis contributed, as defined
+    by total `tx_add_input` values:
+    - MUST transmit their `tx_signatures` before the peer
+  - MUST order the `witness_stack`s by the `serial_id` of the input they
+    correspond to
+  - number of `witness_stack`s MUST equal the number of inputs they added
+
+The receiving node:
+  - MUST fail the negotiation if:
+    - if they have not sent a `tx_signatures` message and they are the
+      peer with the lowest total satoshis contributed, as defined by
+      `tx_add_input` values
+    - the message contains an empty `witness_stack`
+    - the number of `witness_stack`s does not equal the number of inputs
+      added by the sending node
+    - the `txid` does not match the txid of the transaction
+  - SHOULD apply the `witness`es to the transaction and broadcast it
+  - MUST reply with their `tx_signatures` if not already transmitted
+
+#### Rationale
+
+The peer with the lowest total of inputs amounts must transmit its
+`tx_signatures` first. This gives in a strict ordering of transmission
+for any multiparty tx collaboration.
```

Comment:
This isn't completely true, we don't specify yet what happens when both peers contribute the same amount.
I think we should mandate that in that case, the initiator starts?
WDYT @niftynei?

Reply:
Agreed, I think that's a good tie-breaker!

Reply:
FYI simplified updates proposal uses the lexicographical-sorted pubkeys as tie-breaker. Only real difference I can see is initiator doesn't generalize to N party... :shrug:

Reply:
We could also remove this ordering clause and leave it as unspecified? The current implementation just sends the signatures whenever it has them.

I think ordering makes finding the source of a deadlock easier tho, so might be worth adding for that reason.

Reply:
I'd like to keep an ordering, it makes it much easier to fully specify your state machine when implementing.
I don't mind which one we choose, but I'd like one!

> FYI simplified updates proposal uses the lexicographical-sorted pubkeys as tie-breaker. Only real difference I can see is initiator doesn't generalize to N party... :shrug:

Interesting, that's definitely a good option as the interactive-tx protocol should easily generalize to N party transaction building. We can do that on the funding pubkeys, how does that sound @niftynei?

Reply:
> why would a deadlock be expected?

There is an advantage to being the last party to send your sigs, because then you're sure you have a fully signed transaction whenever you send your sigs. Whereas if you send them first, your peer may keep them but never give you their sigs, and you're unable to publish the transaction yourself but still have to remember it until it's double-spent.

Without a specified ordering, there would be an incentive to always send our sigs last, which would lead to a deadlock if both peers apply this strategy.

Reply:
> why would a deadlock be expected? signatures aggregation can proceed in any order converging on the same set of signatures (a G-set in CRDT parlance)

If one party fails to send signatures, the negotiation fails. Ordering them makes it obvious which party failed to send signatures.

Reply:
>  We can do that on the funding pubkeys, how does that sound @niftynei?

Works for me, thanks!